# Análisis de datos "Titánic"

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as mtl

In [2]:
df_gs = pd.read_csv('./titanic/gender_submission.csv')
dF_train = pd.read_csv('./titanic/train.csv')
df_test= pd.read_csv('./titanic/test.csv')

df_gs_copia = df_gs.copy()
df_train_copy = dF_train.copy()
df_test_copy =  df_test.copy()

## Tareas a realizar 

### 🧩 Checklist de tareas — Análisis intermedio (Titanic)
1️⃣ Revisión inicial del dataset

Analizar dimensiones del dataset.

Identificar tipos de variables (numéricas, categóricas, ordinales).

Verificar consistencia de los tipos de datos.

Detectar valores faltantes por columna.

Verificar existencia de registros duplicados.

2️⃣ Calidad de datos y valores faltantes

Evaluar el impacto de los valores nulos en cada variable.

Definir un criterio de tratamiento para:

Edad

Puerto de embarque

Cabina

Justificar cada decisión de imputación, eliminación o transformación.

3️⃣ Consistencia y estandarización

Revisar inconsistencias en variables categóricas.

Unificar formatos de texto.

Validar rangos lógicos en variables numéricas (edad, tarifa, familiares).

Identificar valores imposibles o sospechosos.

4️⃣ Análisis de duplicados

Detectar duplicados exactos.

Evaluar duplicados parciales (mismo nombre, edad, clase).

Definir criterios para conservar o eliminar registros.

5️⃣ Análisis de outliers

Detectar valores atípicos en variables numéricas relevantes.

Evaluar si los outliers son errores o casos válidos.

Decidir el tratamiento adecuado (mantener, limitar o excluir).

Justificar la decisión tomada.

6️⃣ Transformación de variables

Convertir variables a tipos de datos adecuados.

Reorganizar columnas para facilitar el análisis.

Renombrar variables para mayor claridad.

7️⃣ Feature engineering

Crear nuevas variables que aporten información relevante.

Agrupar edades en rangos significativos.

Construir variables derivadas del núcleo familiar.

Generar indicadores binarios útiles para análisis posteriores.

8️⃣ Análisis exploratorio dirigido

Analizar la relación entre supervivencia y variables clave.

Comparar métricas entre grupos relevantes.

Detectar patrones y tendencias significativas.

Identificar variables con mayor poder explicativo.

9️⃣ Validación del dataset final

Verificar ausencia de valores nulos críticos.

Confirmar coherencia entre variables relacionadas.

Revisar que el dataset sea apto para visualización.

Documentar todas las transformaciones realizadas.

🔟 Preparación para Power BI

Definir qué métricas y dimensiones serán utilizadas.

Validar estructura adecuada para segmentadores.

Exportar dataset final en formato compatible.

Verificar integridad de los datos exportados.

## Resolución del análisis

#### 1️⃣ Revisión inicial del dataset

##### Analizar dimensiones del dataset.

In [3]:
df_train_copy.shape

(891, 12)

##### Identificar tipos de variables (numéricas, categóricas, ordinales).


In [4]:
df_train_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


##### Verificar consistencia de los tipos de datos.

###### La unica columna la cual podriamos modificar es "Age" , ya que figura de tipo FLOAT y deberia de ser INT

##### Detectar valores faltantes por columna.

In [5]:
df_train_copy.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

##### Verificar existencia de registros duplicados.

In [6]:
df_train_copy.apply(lambda col: col.duplicated().sum())

PassengerId      0
Survived       889
Pclass         888
Name             0
Sex            889
Age            802
SibSp          884
Parch          884
Ticket         210
Fare           643
Cabin          743
Embarked       887
dtype: int64

#### Conclusión:
La base de datos Train contiene 12 columnas y 891 filas.

Al iniciar el análisis, se revisaron los tipos de variables de las distintas columnas y se detectó una inconsistencia en la variable “age”, la cual será considerada para su corrección.

Al verificar las columnas con datos faltantes, se observó que solo tres presentan valores ausentes; para el desarrollo del análisis se trabajará únicamente con la variable “age”.

Por último, se analizaron los datos duplicados y se identificó que algunas columnas contienen valores numéricos que en realidad corresponden a variables categóricas. Si bien esto no afecta directamente el análisis, se tendrá en cuenta para evitar posibles inconsistencias.

#### 2️⃣ Calidad de datos y valores faltantes

##### 
Visualizamos variables nulas en tres columnas; la más importante con la cual deberíamos comenzar a trabajar es la variable “Age”. Para trabajar sobre la misma, considero que se debería calcular la media del total de las filas de la variable recién nombrada, con el fin de rellenar esos espacios faltantes y poder trabajar correctamente con ella. Respecto a la variable "Embarked" solo tiene dos variables faltantes, podriamos trabajar en ella e intentar completar estas mismas, respecto a la variable cabin , su numero de datos faltantes es muy alto, y la importancia de la misma es poca, decidiria eliminar dicha columna para que no rompa en si nuestro analisis

In [7]:
df_train_copy.groupby('Sex')['Age'].size()

Sex
female    314
male      577
Name: Age, dtype: int64

In [8]:
df_train_copy[df_train_copy['Sex']=='female']['Age'].median()

27.0

In [9]:
df_train_copy[df_train_copy['Sex']=='male']['Age'].median()

29.0

In [10]:
df_train_copy.loc[df_train_copy['Sex'] == 'female','Age'] = df_train_copy.loc[df_train_copy['Sex'] == 'female','Age'].fillna(df_train_copy.loc[df_train_copy['Sex']=='female']['Age'].median())

In [11]:
df_train_copy.loc[df_train_copy['Sex'] == 'male','Age'] = df_train_copy.loc[df_train_copy['Sex'] == 'male','Age'].fillna(df_train_copy[df_train_copy['Sex'] == 'male']['Age'].median())

In [12]:
df_train_copy['Age'] = df_train_copy['Age'].astype(int)

In [13]:
df_train_copy.groupby('Embarked')['Fare'].agg(['min','max'])

min       max
Embarked                  
C         4.0125  512.3292
Q         6.7500   90.0000
S         0.0000  263.0000

In [14]:
df_train_copy['FamilySize'] = df_train_copy['SibSp'] + df_train_copy['Parch'] + 1
df_train_copy['FareForPerson'] = df_train_copy['Fare'] / df_train_copy['FamilySize']

In [15]:
df_train_copy.groupby('Embarked')['FareForPerson'].agg(['max','min'])

max       min
Embarked                    
C         512.3292  2.409733
Q          45.0000  2.583333
S         221.7792  0.000000

In [16]:
df_train_copy.groupby('Embarked')['PassengerId'].size()

Embarked
C    168
Q     77
S    644
Name: PassengerId, dtype: int64

In [17]:
df_train_copy['Embarked'] = df_train_copy['Embarked'].fillna('S')

In [18]:
df_train_copy = df_train_copy.drop(columns = ['Cabin'])

In [19]:
df_train_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PassengerId    891 non-null    int64  
 1   Survived       891 non-null    int64  
 2   Pclass         891 non-null    int64  
 3   Name           891 non-null    object 
 4   Sex            891 non-null    object 
 5   Age            891 non-null    int64  
 6   SibSp          891 non-null    int64  
 7   Parch          891 non-null    int64  
 8   Ticket         891 non-null    object 
 9   Fare           891 non-null    float64
 10  Embarked       891 non-null    object 
 11  FamilySize     891 non-null    int64  
 12  FareForPerson  891 non-null    float64
dtypes: float64(2), int64(7), object(4)
memory usage: 90.6+ KB


#### Conclusión:
En relación con la variable Age, se decidió imputar los valores faltantes utilizando la mediana calculada de forma separada según el género, diferenciando entre pasajeros femeninos y masculinos. Esta decisión se tomó con el objetivo de preservar posibles diferencias en la distribución de edades entre ambos grupos y reducir el impacto de valores atípicos. Posteriormente, se transformó el tipo de dato de la variable a entero (INT), dado que la edad es una variable numérica discreta.

Respecto a la variable Embarked, se observó que presenta únicamente dos valores faltantes sobre un total de 891 registros. Dado su carácter categórico, se analizó la moda de la variable y se identificó que el valor “S” es el más frecuente, representando aproximadamente el 67,18 % de los registros. En función de esto, se decidió imputar los valores nulos con dicho valor.

En cuanto a la variable Cabin, se determinó su eliminación del análisis, ya que presenta un alto porcentaje de valores faltantes (cercano al 70 %), lo que limita su utilidad analítica y podría introducir sesgos en etapas posteriores del análisis

#### 3️⃣ Consistencia y estandarización

In [20]:
df_train_copy.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  891.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.425365    0.523008   
std     257.353842    0.486592    0.836071   13.027360    1.102743   
min       1.000000    0.000000    1.000000    0.000000    0.000000   
25%     223.500000    0.000000    2.000000   22.000000    0.000000   
50%     446.000000    0.000000    3.000000   29.000000    0.000000   
75%     668.500000    1.000000    3.000000   35.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  FamilySize  FareForPerson  
count  891.000000  891.000000  891.000000     891.000000  
mean     0.381594   32.204208    1.904602      19.916375  
std      0.806057   49.693429    1.613459      35.841257  
min      0.000000    0.000000    1.000000       0.000000  
25%      0.000000    7.910400    1.000000       7.250000  
50%      0.000000   14.454200    1.000000       8.300000  
75%      0.000000   31.000000    2.000000      23.666667  
max      6.000000  512.329200   11.000000     512.329200

In [21]:
df_train_copy.nunique()

PassengerId      891
Survived           2
Pclass             3
Name             891
Sex                2
Age               71
SibSp              7
Parch              7
Ticket           681
Fare             248
Embarked           3
FamilySize         9
FareForPerson    289
dtype: int64

#### Conclusión:
Durante esta etapa se verificó la consistencia de los datos y la estandarización de las variables categóricas. No se identificaron valores ilógicos ni diferencias en los formatos de escritura, por lo que no fue necesario realizar correcciones.

#### 4️⃣ Análisis de duplicados

In [22]:
df_train_copy.apply(lambda col: col.duplicated().sum())

PassengerId        0
Survived         889
Pclass           888
Name               0
Sex              889
Age              820
SibSp            884
Parch            884
Ticket           210
Fare             643
Embarked         888
FamilySize       882
FareForPerson    602
dtype: int64

In [23]:
df_train_copy[df_train_copy.duplicated()]

Empty DataFrame
Columns: [PassengerId, Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Embarked, FamilySize, FareForPerson]
Index: []

#### Conclusion : 
El análisis de valores duplicados mostró repeticiones esperables en variables categóricas y discretas. Estas repeticiones no representan errores en los datos, sino un comportamiento normal del dataset, por lo que no fue necesario aplicar correcciones.

#### 5️⃣ Análisis de outliers

In [24]:
df_train_copy[df_train_copy['Fare'] == 0]

PassengerId  Survived  Pclass                              Name   Sex  \
179          180         0       3               Leonard, Mr. Lionel  male   
263          264         0       1             Harrison, Mr. William  male   
271          272         1       3      Tornquist, Mr. William Henry  male   
277          278         0       2       Parkes, Mr. Francis "Frank"  male   
302          303         0       3   Johnson, Mr. William Cahoone Jr  male   
413          414         0       2    Cunningham, Mr. Alfred Fleming  male   
466          467         0       2             Campbell, Mr. William  male   
481          482         0       2  Frost, Mr. Anthony Wood "Archie"  male   
597          598         0       3               Johnson, Mr. Alfred  male   
633          634         0       1     Parr, Mr. William Henry Marsh  male   
674          675         0       2        Watson, Mr. Ennis Hastings  male   
732          733         0       2              Knight, Mr. Robert J  male   
806          807         0       1            Andrews, Mr. Thomas Jr  male   
815          816         0       1                  Fry, Mr. Richard  male   
822          823         0       1   Reuchlin, Jonkheer. John George  male   

     Age  SibSp  Parch  Ticket  Fare Embarked  FamilySize  FareForPerson  
179   36      0      0    LINE   0.0        S           1            0.0  
263   40      0      0  112059   0.0        S           1            0.0  
271   25      0      0    LINE   0.0        S           1            0.0  
277   29      0      0  239853   0.0        S           1            0.0  
302   19      0      0    LINE   0.0        S           1            0.0  
413   29      0      0  239853   0.0        S           1            0.0  
466   29      0      0  239853   0.0        S           1            0.0  
481   29      0      0  239854   0.0        S           1            0.0  
597   49      0      0    LINE   0.0        S           1            0.0  
633   29      0      0  112052   0.0        S           1            0.0  
674   29      0      0  239856   0.0        S           1            0.0  
732   29      0      0  239855   0.0        S           1            0.0  
806   39      0      0  112050   0.0        S           1            0.0  
815   29      0      0  112058   0.0        S           1            0.0  
822   38      0      0   19972   0.0        S           1            0.0

#### Conclusión :
No se visualizaron valores atípicos relevantes en el conjunto de datos, por lo que no fue necesario aplicar correcciones adicionales

#### 6️⃣ Transformación de variables

In [25]:
df_train_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PassengerId    891 non-null    int64  
 1   Survived       891 non-null    int64  
 2   Pclass         891 non-null    int64  
 3   Name           891 non-null    object 
 4   Sex            891 non-null    object 
 5   Age            891 non-null    int64  
 6   SibSp          891 non-null    int64  
 7   Parch          891 non-null    int64  
 8   Ticket         891 non-null    object 
 9   Fare           891 non-null    float64
 10  Embarked       891 non-null    object 
 11  FamilySize     891 non-null    int64  
 12  FareForPerson  891 non-null    float64
dtypes: float64(2), int64(7), object(4)
memory usage: 90.6+ KB


#### Conclusión :
En la etapa de transformación de variables, se realizó la conversión del tipo de dato de la variable “Age”, pasando de float a int, con el objetivo de mantener coherencia con su naturaleza numérica. No se identificó la necesidad de aplicar transformaciones adicionales en el resto de las variables

#### 7️⃣ Feature engineering

#### Conclusión :
Durante la etapa de feature engineering se crearon variables derivadas relacionadas con el núcleo familiar y el costo por persona, con el objetivo de enriquecer el análisis. Estas nuevas variables permiten una mejor interpretación del comportamiento de los pasajeros.

#### 8️⃣ Análisis exploratorio dirigido

In [26]:
df_train_copy.groupby('Sex')['Survived'].mean()*100

Sex
female    74.203822
male      18.890815
Name: Survived, dtype: float64

##### Podemos visualizar que el 74,20% de las mujeres , y el 18,89% de los hombres sobrevivió

In [27]:
df_train_copy.groupby(['Sex','Pclass'])['Survived'].mean().unstack()

Pclass         1         2         3
Sex                                 
female  0.968085  0.921053  0.500000
male    0.368852  0.157407  0.135447

bins = [0,15,30,45,100]
labels = ['0-15','15-30','30-45','45+']

(df_train_copy.groupby([pd.cut(df_train_copy['Age'],bins = bins , labels = labels, right = False),'Sex'])['Survived'].mean()*100).round(1)

#### Conclución :
El análisis muestra que la supervivencia estuvo fuertemente influenciada por el género. Si bien en el grupo de niños las tasas son similares entre hombres y mujeres, en los rangos de edad mayores las mujeres presentan consistentemente mayores probabilidades de supervivencia, evidenciando una clara prioridad hacia este grupo.

#### Sistema de guardado e importacion

In [28]:
import os
os.chdir(r"C:\Users\ASUS\Desktop")

In [29]:
os.listdir()

['anaconda-navigator.exe - Acceso directo.lnk',
 'Block notas Necesarias',
 'DBeaver.lnk',
 'desktop.ini',
 'Docker Desktop.lnk',
 'EspinozaRodrigo-TP.docx',
 'GitHub Desktop.lnk',
 'Google Chrome.lnk',
 'Isa',
 'Microsoft Edge.lnk',
 'MSSQL16.SQLEXPRESS01',
 'MySQL Workbench 8.0 CE.lnk',
 'Python 3.13 (64-bit).lnk',
 'python nuevo',
 'respaldo.reg',
 'Roblox Player.lnk',
 'Rodrigo_Espinoza_CV.pdf',
 'Software Fix.lnk',
 'titulos ontenidos',
 'UTN',
 'Visual Studio Code.lnk']

In [30]:
print(os.listdir("C://Users"))

['.kaggle', 'All Users', 'ASUS', 'Default', 'Default User', 'desktop.ini', 'Public']


In [33]:
df_train_copy.to_csv('titanic_limpiezaCompleta.csv', index=False)

In [34]:
import os
os.listdir()

['anaconda-navigator.exe - Acceso directo.lnk',
 'Block notas Necesarias',
 'DBeaver.lnk',
 'desktop.ini',
 'Docker Desktop.lnk',
 'EspinozaRodrigo-TP.docx',
 'GitHub Desktop.lnk',
 'Google Chrome.lnk',
 'Isa',
 'Microsoft Edge.lnk',
 'MSSQL16.SQLEXPRESS01',
 'MySQL Workbench 8.0 CE.lnk',
 'Python 3.13 (64-bit).lnk',
 'python nuevo',
 'respaldo.reg',
 'Roblox Player.lnk',
 'Rodrigo_Espinoza_CV.pdf',
 'Software Fix.lnk',
 'titanic_limpiezaCompleta.csv',
 'titulos ontenidos',
 'UTN',
 'Visual Studio Code.lnk']

In [35]:
import os
os.getcwd()

'C:\\Users\\ASUS\\Desktop'